theano/compile/builders.py

```diff
@@ -115,40 +73,73 @@ def __hash__(self):
         # TODO: use internal variables in hash
         return hash(type(self))
 
+    def grad(self, inputs, output_grads):
+        if self.grad_op_is_cached:
+            return self.grad_op(inputs, output_grads)
+
+        if self.grad_op is None:
+            self.grad_op = []
+
+        # we need to convert a list into a single funtor
+        if isinstance(self.grad_op, list):
+            grad_op_l = self.grad_op
+            if len(grad_op_l) > len(self.local_inputs):
+                raise ValueError(
+                    'Can override %d gradients at most, got %d' % (
+                        len(self.local_inputs), len(grad_op_l)))
+            if len(grad_op_l) < len(self.local_inputs):
+                grad_op_l += [None] * (
+                    len(self.local_inputs) - len(grad_op_l))
+            wrt = [self.local_inputs[i] for i, go in
+                   enumerate(grad_op_l) if not go]
+            # compute non-overriding downsteam gradients from upstreams grads
+            # it's normal some input may be disconnected, thus the 'ignore'
+            ups_grads_d = dict(izip(self.local_outputs, output_grads))
+            nat_dns_grads = iter(theano.gradient.grad(
+                cost=None,
+                known_grads=ups_grads_d,
+                wrt=wrt,
+                disconnected_inputs='ignore'))
```

Comment:
Again, not so sure about the variable names, `ups_grads_d` and `nat_dns_grads` are quite cryptic. Maybe specify `known_grads=` directly inline, and call the other variable `default_grads` or something, because they are not overridden?
Code-wise, it'd be good to wrap this in `if wrt:`, so you avoid the `grad` call if all gradients are overridden anyway. You can even leave `nat_dns_grads` undefined if `wrt` is empty, or set it to `iter([])` in the `else:` case.